Implicit contract

fubuloubu opened the pull request:
### What was wrong?
Wanted a contract type that could magically figure out (from the ABI) when a method should be a call by default, or a transaction by default

### How was it fixed?
I added a subclass of `ConciseContract` that checked the ABI whether or not the method was `constant`/`pure` ('constant': True in ABI), and passed that info to a subclass of `ConciseMethod`. If it was determined to make a transaction by default, the transaction is made using the defaultAccount and zero value.

#### Cute Animal Picture
![Cute animal picture](http://www.lifenews.com/wp-content/uploads/2014/12/orangutan.jpg)

pipermerriam commented on requirements-dev.txt:
```diff
@@ -4,6 +4,7 @@ ethereum>=1.6.1,<2.0
 flaky>=3.3.0
 flake8==3.4.1
 hypothesis>=3.31.2
+lru-dict>=1.1.6
```
this requirement is already present in `setup.py`.  You just need to re-install the package locally to pickup the new requirement.

fubuloubu replied:
I'll remove this commit, glad I made it separately.

pipermerriam commented on web3/contract.py:
```diff
@@ -764,9 +765,48 @@ def __prepared_function(self, **kwargs):
+    '''
+    def __getattr__(self, attr):
+        # If method_name.constant is true in the abi, then call by default
+        call_by_default = \
```
This is too difficult to parse/read.  I would recommend breaking this out into a small utility function which **should** also make it easier to test the abi inspection logic directly.

Alternatively you could break this up some across multiple lines, but I think this would benefit from being a stand-alone easy-to-test function.

fubuloubu replied:
`_find_matching_fn_abi` makes this cleaner. I separated it out

pipermerriam commented on web3/contract.py:
```diff
@@ -764,9 +765,48 @@ def __prepared_function(self, **kwargs):
+        # Defaults are: from coinbase and no value sent
+        if not kwargs and not self.call_by_default:
+            default_account = self._contract.web3.eth.defaultAccount
+            return super()._prepared_function(transact={"from": default_account, "value": 0})
```
I think that `value` defaults to zero if not included in the transaction dict which suggests it shouldn't be included here.

fubuloubu replied:
:+1: so just `from` is needed?

pipermerriam commented on web3/contract.py:
```diff
@@ -764,9 +765,48 @@ def __prepared_function(self, **kwargs):
+        call_by_default = \
+                [m['constant'] if 'constant' in m.keys() else False \
+                    for m in self._classic_contract.abi if \
+                        m['type'] != 'constructor' and m['name'] == attr][0]
```
There is an api on the contract object `_find_matching_fn_abi` that will retrieve the ABI for a named function for you.

fubuloubu replied:
:+1: will update, also helps simplify

fubuloubu commented:
As a part of #419 will the tests for `ConciseContract` be updated?

I was also just thinking I only added two tests that are unique to this subclass, perhaps I can forgo the original `ConciseContract` tests, keeping `test_implicitcontract_transact_default` and `test_implicitcontract_abi_inspection` only. The first fully tests (IMO) the functionality behind `ImplicitMethod` even though `Mock` is used, and the second fully tests that the correct calls are made on a real contract. I basically set up the flag to check whether or not the method should be called by default when the call is made with the given context, I don't see how the `Mock` really plays into this.

fubuloubu commented:
Should also keep `test_implicitcontract_call_default` as this shows the other side of the coin for `ImplicitMethod`

pipermerriam commented:
I'm still not ok with the mock based tests.  They represent an anti-pattern and I can't accept the maintenance cost they represent.  A few end-to-end tests which use real contracts and a real EVM should be sufficient to test the code.  I don't think we necessarily need to directly test the `ImplicitMethod` object.

fubuloubu commented:
Alright then, will do!

Does this mean `ConciseContract`'s should get updated as well for the same concern?

fubuloubu commented:
Changed all the test cases to check the specific functionality I added in my subclass. I am no longer testing any functionality in the `ConciseContract`/`ConciseMethod` classes.

I couldn't think of a way to test it by asking if the contract had made a `transact`/`call` (like how `ConciseContract` is tested when using `unittest.Mock`), so instead I simply checked if the transaction was added (and automined) to the latest block.

If this is not sufficient, please let me know.

fubuloubu commented:
Noticed there was a pretty large rewrite of `web3/contracts.py` so I had to completely redo my commits here. Squashed down to one commit.

carver commented on web3/contract.py:
```diff
@@ -698,17 +698,17 @@ def _none_addr(datatype, data):
 
 
-class ConciseMethod:
+class ConciseMethod(object):
```
This shouldn't be necessary, web3 is py3-only.

fubuloubu replied:
Does this happen under the hood for py3? I always thought `class MyObj` was py2-preferred and `class MyObj(object)` was py3

carver replied:
> I always thought class MyObj was py2-preferred and class MyObj(object) was py3

It is the reverse, actually. `class MyObj(object)` was how you got some Py3-style object features into Py2. It is [(nearly) a no-op in Py3](https://stackoverflow.com/questions/1238606/is-it-necessary-or-useful-to-inherit-from-pythons-object-in-python-3-x).

fubuloubu replied:
Ah, you learn something new everyday. Or more accurately, unlearn a bad habit lol.

fubuloubu replied:
P.S. There are 26 instances of this in `web3/`:
```bash
$ grep -r "(object)" web3/ | wc -l
26
```

carver replied:
Yeah, not a surprise that some are leftover from our py2 support days. Do you want to do the honors of deleting them? :)

If not, I'll take care of it.

carver commented on web3/contract.py:
```diff
@@ -698,17 +698,17 @@ def _none_addr(datatype, data):
+        return self._prepared_function(*args, **kwargs)
 
-    def __prepared_function(self, *args, **kwargs):
+    def _prepared_function(self, *args, **kwargs):
```
I'm kind of paranoid about causing name conflicts, and this exposes us to another one. Some options are:

1. Rename to `_web3py_prepared_function` which would be a diabolical thing to call a Solidity function :P
1. Extract to a utility function, removing it from the class entirely, and pass in `self.__function` as an argument.
2. Rename to a Solidity keyword that makes sense, although the only one I can think of is `function` which is confusingly close to `__function` and the classic contract's `functions`.

I have a slight preference for 1, mostly because 2 might be a big rewrite. I have a strong preference for 2 over 3 (but could be convinced if the perfect keyword comes along).

fubuloubu replied:
I could create *another* method in this class called something like `_get_prepared_function(*args, **kwargs)`, that calls this directly. Trying to figure out why this function needs to be totally private versus protected. Or I can just completely go `super()._ConciseContract__prepared_function(*args, **kwargs)` to call it in the subclass. Obviously not preferred, but if the shoe fits...

fubuloubu replied:
Figured out a workaround

fubuloubu commented on tests/core/contracts/test_implicit_contract.py:
```diff
@@ -0,0 +1,89 @@
+    blocknum, starting_txns = get_transaction_count("pending")
+    # When a function is called that defaults to call
+    # (Verify something was returned)
+    assert math_contract.counter() is not None
```
Can do

fubuloubu replied:
1/4 tests covers this concern. If it works for one call, it works for any call to the same method I'd say.

fubuloubu commented on tests/core/contracts/test_implicit_contract.py:
```diff
@@ -0,0 +1,89 @@
+    assert math_contract.return13() is not None
+    # Check that a transaction was made and not a call
+    # (Auto-mining is enabled, so query by block number)
+    assert get_transaction_count(blocknum) == starting_txns + 1
```
I like it, will do

fubuloubu replied:
I don't know `return13()` returns, but it certainly isn't the number 13 lol. I don't know where the source code is for this, what return value do you want me to check of this pre-compiled contract?

fubuloubu replied:
Ah, I see how this is different. Disregard previous comment

fubuloubu replied:
I covered all of these concerns, but not in every single test since 3/4 tests do transactions, which returns a txn_hash instead of the actual return value.

fubuloubu commented:
@carver See my individual comments on your prior review. I rewrote what I had initially, reducing the amount of state being stored in my subclass and optimizing around the `_function` member of `ConciseMethod` as a work around in order to find whether or not the method being called is `constant` in the abi.

carver commented:
Yup, that all looks good to me!

It looks like you addressed @pipermerriam 's comments, too, but I'll wait for the rest of the day for him to review again if he likes. If he's too busy, I'll just merge it at end of day.

pipermerriam commented:
Blindly approving.  Blaming it on @carver if anything is wrong :P

fubuloubu commented:
Yay blind approvals!
